File: toolkit/echo_server/echo.py
#!/usr/bin/python
# $Id: echo.python,v 1.7 2001/05/06 22:21:26 doug Exp $
# see: http://www.bagley.org/~doug/shootout/
# with help from Brad Knotwell
# usage: $0 [bytes]
# note: run on unix/linux
# date: 2017/9/6

import sys
import os
from socket import *

DATA = "Hello there sailor\n"
bufferSize = len(DATA)


def server_sock():
    sock = socket(AF_INET, SOCK_STREAM)
    sock.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
    sock.bind(('127.0.0.1', 0))
    sock.listen(2)
    return(sock)


def get_port(sock):
    host, port = sock.getsockname()
    return(port)


def client_sock(port):
    sock = socket(AF_INET, SOCK_STREAM)
    sock.connect(('127.0.0.1', port))
    return(sock)


def echo_client(n, port):
    sock = client_sock(port)
    sender, receiver = sock.send, sock.recv
    for i in range(0, n):
        sender(DATA)
        ans = receiver(bufferSize)
        while ans[-1] != "\n":
            ans += receiver(bufferSize - len(ans))
            if ans <> DATA:
                raise("client: \"%s\" ne \"%s\"" % (DATA, ans))
                sock.close()


def echo_server(n):
    ssock = server_sock()
    if os.fork() > 0:
        # parent is server
        csock, addr = ssock.accept()
        n = 0
        sender, receiver = csock.send, csock.recv
        while 1:
            dat = receiver(bufferSize)
            if not dat:
                break
            sender(dat)
            n += len(dat)
            print "server processed %d bytes" % n
            os.wait()
    else:
        # child is client
        echo_client(n, get_port(ssock))


def main():
    n = int(sys.argv[1])
    if n < 1:
        n = 1
    echo_server(n)


if __name__ == "__main__":
    main()

File: toolkit/echo_server/echo.c
/* -*- mode: c -*-
 * $Id: echo.gcc,v 1.9 2001/06/28 03:04:47 doug Exp $
 * http://www.bagley.org/~doug/shootout/
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <netinet/in.h>


typedef int (*SOCKACTION_P)(int,struct sockaddr *,socklen_t);
#define DATA "Hello there sailor\n"

void myabort (char *m) {
    fprintf(stderr, "%s\n", m);
    exit(1);
}
void sysabort (char *m) {
    perror(m);
    exit(1);
}

int sigchld = 0;
void reaper (int sig) {
    sigchld = 1;
}

int
genericSock(int port,SOCKACTION_P action,char *actionExceptionText) {
    int ss, optval = 1;
    struct sockaddr_in sin;
    if ((ss = socket(PF_INET, SOCK_STREAM, 0)) == -1)
        sysabort("socket");
    if (setsockopt(ss, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)) == -1)
        sysabort("setsockopt");
    memset(&sin,0,sizeof(sin));
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    sin.sin_port = port;
    if (action(ss, (struct sockaddr *)&sin,(socklen_t)sizeof(sin)) == -1)
        sysabort(actionExceptionText);

    return(ss);
}

int
server_sock () {
    int ss = genericSock(0,(SOCKACTION_P)bind,"server/bind");
    return(listen(ss,2),ss);
}

int
client_sock (int port) {
    return(genericSock(port,(SOCKACTION_P)connect,"client/connect"));
}

int
get_port (int sock) {
    struct sockaddr_in sin;
    socklen_t slen = sizeof(sin);
    if (getsockname(sock, (struct sockaddr *)&sin, &slen) == -1)
        sysabort("server/getsockname");
    return(sin.sin_port);
}

void
echo_client (int n, int port) {
    int i, sock, olen, len, nwritten, nread;
    char *offset, obuf[64], ibuf[64];
    char *end = ibuf + sizeof(ibuf);

    sock = client_sock(port);
    strcpy(obuf, DATA);
    olen = strlen(obuf);
    for (i=0; i<n; i++) {
        len = olen;
        offset = obuf;
        while (len > 0) {
            if ((nwritten = write(sock, offset, len)) == -1)
                sysabort("client/write");
            offset += nwritten;
            len -= nwritten;
        }
        offset = ibuf;
        while ((nread = read(sock, offset, (end - offset))) > 0) {
            offset += nread;
            if (*(offset-1) == '\n') break;
        }
        if (nread == -1)
            sysabort("client/read");
        *offset = 0;
        if ((strcmp(obuf, ibuf)) != 0) {
            char mbuf[128];
            sprintf(mbuf, "client: \"%s\" ne \"%s\"", obuf, ibuf);
            myabort(mbuf);
        }
    }
    close(sock);
}

void
echo_server (int n) {
    int ssock, csock, len, nwritten, total_bytes;
    pid_t pid;
    char buf[64], *offset;
    struct sockaddr_in sin;
    socklen_t slen = sizeof(sin);
    int status;

    ssock = server_sock();
    signal(SIGCHLD, reaper);
    if ((pid = fork()) == -1)
        sysabort("server/fork");
    if (pid) {

        if ((csock = accept(ssock, (struct sockaddr *)&sin, &slen)) == -1)
            sysabort("server/accept");
        total_bytes = 0;
        while ((len = read(csock, buf, sizeof(buf))) > 0) {
            if (sigchld) myabort("server/sigchld");
            offset = buf;
            total_bytes += len;
            while (len > 0) {
                if ((nwritten = write(csock, offset, len)) == -1)
                    sysabort("server/write");
                offset += nwritten;
                len -= nwritten;
            }
        }
        if (len == -1)
            sysabort("server/read");
        close(csock);
        fprintf(stdout, "server processed %d bytes\n", total_bytes);
    } else {

        echo_client(n, get_port(ssock));
    }
    wait(&status);
}

int
main(int argc, char *argv[]) {
    echo_server((argc == 2) ? atoi(argv[1]) : 1);
    return(0);
}

